### PR TITLE
Add organization discovery management related endpoints

### PR DESCRIPTION
**Pull request by dewniMW**
Issue https://github.com/wso2/product-is/issues/16379

### CODE REVIEW

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java [hunk 1]
```diff
@@ -662,4 +946,9 @@ private OrgApplicationManager getOrgApplicationManager() {
 
         return OrganizationManagementServiceHolder.getInstance().getOrgApplicationManager();
     }
+
+    private OrganizationDiscoveryManager getOrganizationDiscoveryManager() {
+
+        return OrganizationManagementServiceHolder.getInstance().getOrganizationDiscoveryManager();
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java:952]: move this to a class variable.

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java [hunk 2]
```diff
@@ -653,6 +830,113 @@ private SharedApplicationsResponse createSharedApplicationsResponse(List<SharedA
...
+                    ERROR_CODE_FILTERING_NOT_IMPLEMENTED.getMessage(),
+                    ERROR_CODE_FILTERING_NOT_IMPLEMENTED.getDescription());
+            throw new OrganizationManagementEndpointException(Response.Status.NOT_IMPLEMENTED, error);
+        }
+    }
+
+    private OrganizationMetadata getOrganizationMetadataResponse(Organization organization, List<OrgDiscoveryAttribute>
+            organizationDiscoveryAttributes) {
+
+        OrganizationMetadata organizationMetadata = new OrganizationMetadata();
```
Comment (thanujalk) [inline components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java:895]: Can use builder pattern for this

Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java:895]: OrganizationMetadata is an autogenerated class. hence unable to modify the class to support the builder pattern

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java [hunk 3]
```diff
@@ -79,4 +81,24 @@ public void setOrganizationManager(OrganizationManager organizationManager) {
...
     }
+
+    /**
+     * Get OrganizationDiscoveryManager OSGi service.
+     *
+     * @return OrganizationDiscoveryManager.
+     */
+    public OrganizationDiscoveryManager getOrganizationDiscoveryManager() {
+
+        return OrganizationManagementServiceHolder.getInstance().organizationDiscoveryManager;
```
[the remark this answers is not in the document]
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java:92]: performing this change will result in a null object when calling the method.

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/util/OrganizationManagementEndpointUtil.java [hunk 4]
```diff
@@ -149,6 +150,18 @@ public static URI getResourceLocation(String organizationId) {
                 + organizationId);
     }
 
+    /**
+     * Get location of the created organization discovery attributes.
+     *
+     * @param organizationId The unique identifier of the organization.
+     * @return URI
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/util/OrganizationManagementEndpointUtil.java:157]:
```suggestion
     * @return URI.
```

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml [hunk 5]
```diff
@@ -522,6 +522,251 @@ paths:
...
+        '204':
+          description: Successfully deleted
+        '400':
+          $ref: '#/components/responses/BadRequest'
+        '401':
+          $ref: '#/components/responses/Unauthorized'
+        '403':
+          $ref: '#/components/responses/Forbidden'
+        '404':
+          $ref: '#/components/responses/NotFound'
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml:665]: do we need 404 status code, due to idempotent law

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml [hunk 6]
```diff
@@ -906,6 +1160,159 @@ components:
...
+            $ref: '#/components/schemas/Link'
+          example:
+            [
+              {
+                "href": "/o/10084a8d-113f-4211-a0d5-efe36b082211/api/server/v1/organizations/discovery?filter=attributes.type+eq+emailDomain&limit=10&offset=50",
+                "rel": "next",
+              },  {
+              "href": "/o/10084a8d-113f-4211-a0d5-efe36b082211/api/server/v1/organizations/discovery?filter=attributes.type+eq+emailDomain&limit=10&offset=30",
+              "rel": "previous",
+            }
```
Comment (AnuradhaSK) [inline components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml:1225]: do we provide such data in other index based paginated resources ?

Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml:1225]: yes. ex: https://github.com/wso2/identity-api-server/blob/be5e9ca66df8ff260ab9bd5463d7e7eafcbbd767/components/org.wso2.carbon.identity.api.server.idp/org.wso2.carbon.identity.api.server.idp.v1/src/main/resources/idp.yaml#L2599-L2612

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml (hunk 5 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/resources/org.wso2.carbon.identity.organization.management.yaml]: removed

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java (hunk 1 above)
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.v1/src/main/java/org/wso2/carbon/identity/api/server/organization/management/v1/service/OrganizationManagementService.java:952]: As discussed offline, will be keeping this as it is

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java [hunk 7]
```diff
@@ -79,4 +81,24 @@ public void setOrganizationManager(OrganizationManager organizationManager) {
...
+    }
+
+    /**
+     * Set OrganizationDiscoveryManager OSGi service.
+     *
+     * @param organizationDiscoveryManager OrganizationDiscoveryManager.
+     */
+    public void setOrganizationDiscoveryManager(OrganizationDiscoveryManager organizationDiscoveryManager) {
+
+        OrganizationManagementServiceHolder.getInstance().organizationDiscoveryManager = organizationDiscoveryManager;
```
[the remark this answers is not in the document]
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java:102]: As discussed offline, will be keeping this as it is

File: components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java (hunk 3 above)
[the remark this answers is not in the document]
Comment (dewniMW) [reply components/org.wso2.carbon.identity.api.server.organization.management/org.wso2.carbon.identity.api.server.organization.management.common/src/main/java/org/wso2/carbon/identity/api/server/organization/management/common/OrganizationManagementServiceHolder.java:92]: As discussed offline, will be keeping this as it is

**Review by thanujalk — APPROVED** [review]

**Review by AnuradhaSK — APPROVED** [review]

**Comment by dewniMW** [comment]
The PR builder has an issue with applying the diff. I've run the relevant integration tests locally to confirm that there are no failures. Hence merging the PR

**Pull request merged**